lib/gringotts/gateways/sagepay.ex

```diff
@@ -0,0 +1,362 @@
+  def refund(amount, payment_id, opts) do
+    # commit(args, ...)
+  end
+
+  @doc """
+  Stores the payment-source data for later use.
+
+  > This usually enable "One Click" and/or "Recurring Payments"
+
+  ## Note
+
+  > If there's anything noteworthy about this operation, it comes here.
+
+  ## Example
+
+  > A barebones example using the bindings you've suggested in the `moduledoc`.
+  """
+  @spec store(CreditCard.t(), keyword) :: {:ok | :error, Response}
+  def store(%CreditCard{} = card, opts) do
+    # commit(args, ...)
+  end
+
+  @doc """
+  Removes card or payment info that was previously `store/2`d
+
+  Deletes previously stored payment-source data.
+
+  ## Note
+
+  > If there's anything noteworthy about this operation, it comes here.
+
+  ## Example
+
+  > A barebones example using the bindings you've suggested in the `moduledoc`.
+  """
+  @spec unstore(String.t(), keyword) :: {:ok | :error, Response}
+  def unstore(registration_id, opts) do
+    # commit(args, ...)
+  end
+
+  ###############################################################################
+  #                                PRIVATE METHODS                              #
+  ###############################################################################
+
+  # Makes the request to sagepay's network.
+  # For consistency with other gateway implementations, make your (final)
+  # network request in here, and parse it using another private method called
+  # `respond`.
+
+  # @spec commit(_) :: {:ok | :error, Response}
+  def commit(:post, endpoint, params, opts) do
+    a_url = @url <> endpoint
+    response = HTTPoison.post(a_url, params, opts)
+    format_response(response)
+  end
+
+  # Function generate_mkey generate a merchant_session_key that will exist only for 400 seconds
+  # and for 3 wrong card identifiers
+
+  defp generate_mkey(opts) do
+    vender_name = Poison.encode!(%{vendorName: opts[:config].vendor})
+
+    merchant_header = [
+      {"Authorization", "Basic " <> opts[:config].auth_id},
+      {"Content-type", "application/json"}
+    ]
+
+    key =
+      commit(:post, "merchant-session-keys", vender_name, merchant_header)
+      |> Map.get("merchantSessionKey")
+  end
+
+  # Function generate_ckey generate a unique cardidentifier for every transaction
+
+  defp generate_ckey(card, merchant_key) do
+    card_header = [
+      {"Authorization", "Bearer " <> merchant_key},
+      {"Content-type", "application/json"}
+    ]
+
+    card = card |> Poison.encode!()
+
+    card_identifier =
+      commit(:post, "card-identifiers", card, card_header)
+      |> Map.get("cardIdentifier")
+  end
+
+  # Function card_details creates the actual body (details of the customer )of the card 
+  # and with merchant_session_key, card_identifiier ,shipping address of a
+  # customer, and other details and converting the map into keyword list
+
+  defp card_details(amount, merchant_key, card_key, opts) do
```

Comment:
This function does a lot more than just build card details. Would be better to split it into _at least_ `card_details` and `other_details`.